### PR TITLE
Bug 1120358 - Part2: Gaia JS integration test case

### PR DESCRIPTION
**Pull request by chenpighead**
Bug 1120358 - Part2: Gaia JS integration test case

### CODE REVIEW

**Comment by try-server-hook** [comment]
[chenpighead Chun-Min (Jeremy) Chen (chenpighead)](https://github.com/chenpighead) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e5e973faacbc)

File: apps/system/test/marionette/lib/faketextselectionapp.js [hunk 1]
```diff
@@ -120,6 +126,13 @@ FakeTextSelectionApp.prototype = {
     this.textSelection.longPress(dom);
   },
 
+  longPressByPosition: function(target,x,y) {
```
Comment (cctuan) [inline apps/system/test/marionette/lib/faketextselectionapp.js]: (target, x, y)

File: apps/system/test/marionette/lib/text_selection.js [hunk 2]
```diff
@@ -88,6 +88,13 @@ TextSelection.prototype = {
       press(element, eltSize.width / 2, eltSize.height / 2).
       moveByOffset(0, 0).wait(2).release().perform();
   },
+  
+  longPressByPosition: function(element,x,y) {
```
Comment (cctuan) [inline apps/system/test/marionette/lib/text_selection.js]: (element, x, y)

File: apps/system/test/marionette/lib/text_selection.js [hunk 3]
```diff
@@ -88,6 +88,13 @@ TextSelection.prototype = {
...
+    // Add moveByOffset to prevent contextmenu event be fired.
+    var eltSize = element.size();
```
Comment (cctuan) [inline apps/system/test/marionette/lib/text_selection.js]: no need

File: apps/system/test/marionette/lib/text_selection.js [hunk 4]
```diff
@@ -88,6 +88,13 @@ TextSelection.prototype = {
...
+    this.actions.tap(element, x, y).wait(2).press(element, x, y).
+      moveByOffset(0, 0).wait(2).release().perform();
+  },
```
Comment (cctuan) [inline apps/system/test/marionette/lib/text_selection.js]: Could you also modify longPress method?
```
longPress: function(element) {
  var eltSize = element.size();
  this.longPressByPosition(element, eltSize/2, eltSize/2);
}
```

File: apps/system/test/marionette/text_selection_test.js [hunk 5]
```diff
@@ -265,6 +265,33 @@ marionette('Text selection >', function() {
...
     });
+
+    suite('selection carets bug', function() {
+      setup(function() {
+        fakeTextselectionApp.setTestFrame('bug1120358');
+      });
+      test('bug1120358 : Positions of selection carets should be updated ' +
+           'correctly while scrolling (so far on device only)',
+        function() {
+          fakeTextselectionApp.longPressByPosition('BugContent',30,150);
```
Comment (cctuan) [inline apps/system/test/marionette/text_selection_test.js]: ..., 30, 150)

File: apps/system/test/marionette/text_selection_test.js [hunk 6]
```diff
@@ -265,6 +265,33 @@ marionette('Text selection >', function() {
...
+                               'scrollTop += 20');
+          client.helper.wait(500);
+          var flickPosition = {};
+          flickPosition.caretB = {
+            offset: {
+              x: 20, y: 0
+            }
+          };
+          fakeTextselectionApp.BugContent.selectionHelper.
+          moveCaretByPosition(flickPosition);
```
Comment (cctuan) [inline apps/system/test/marionette/text_selection_test.js]: just use 
```
moveCaretByPosition({
  caretB: {
    offset: {x: 20, y:0}
  }
})
```

File: apps/system/test/marionette/text_selection_test.js [hunk 7]
```diff
@@ -265,6 +265,33 @@ marionette('Text selection >', function() {
...
+          client.helper.wait(500);
+          assert.ok(fakeTextselectionApp.BugContent.selectionHelper.
+          selectedContent.length > 0,
```
Comment (cctuan) [inline apps/system/test/marionette/text_selection_test.js]: I think we can compare the selectedContent before/after scrolling. What do you think?

Comment (chenpighead) [inline apps/system/test/marionette/text_selection_test.js]: Good idea! I've recommited a new patch to change the test assertion criterion.

File: apps/system/test/marionette/text_selection_test.js (hunk 6 above)
Comment (chenpighead) [inline apps/system/test/marionette/text_selection_test.js]: Since we change the test assertion criterion, yhis will be changed as well.

**Comment by try-server-hook** [comment]
[chenpighead Chun-Min (Jeremy) Chen (chenpighead)](https://github.com/chenpighead) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=bc706cf80d7b)

**Comment by try-server-hook** [comment]
[chenpighead Chun-Min (Jeremy) Chen (chenpighead)](https://github.com/chenpighead) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6e227be53029)

**Comment by try-server-hook** [comment]
[chenpighead Chun-Min (Jeremy) Chen (chenpighead)](https://github.com/chenpighead) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=a88c8ee9141c)

File: apps/system/test/marionette/lib/text_selection.js [hunk 8]
```diff
@@ -82,10 +82,13 @@ TextSelection.prototype = {
   },
 
   longPress: function(element) {
-    // Add moveByOffset to prevent contextmenu event be fired.
     var eltSize = element.size();
-    this.actions.tap(element, eltSize.width / 2, eltSize.height / 2).wait(2).
-      press(element, eltSize.width / 2, eltSize.height / 2).
+    this.longPressByPosition(element, eltSize/2, eltSize/2);
+  },
+
```
Comment (cctuan) [inline apps/system/test/marionette/lib/text_selection.js]: trailing whitespace.

**Comment by try-server-hook** [comment]
[chenpighead Chun-Min (Jeremy) Chen (chenpighead)](https://github.com/chenpighead) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=233f6753d63b)

File: apps/system/test/marionette/lib/text_selection.js (hunk 8 above)
Comment (chenpighead) [inline apps/system/test/marionette/lib/text_selection.js]: I've removed it. thx!

**Comment by try-server-hook** [comment]
[chenpighead Chun-Min (Jeremy) Chen (chenpighead)](https://github.com/chenpighead) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=5b8c3dc6831b)

**Comment by cctuan** [comment]
try to reopen it to trigger tbpl.

**Comment by try-server-hook** [comment]
[cctuan George (cctuan)](https://github.com/cctuan) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=01d160e85fe3)

**Pull request merged**